### PR TITLE
`SpriteRenderer` add tiled `DrawMode`

### PR DESCRIPTION
**Pull request by cptbtptpbcptdtptp**
When scaling sprites, developers want to set the `DrawMode` style according to their own intentions, now we provide the following three modes:
- Simple
merge
- Slice
merge
- Tiled
merge

### CODE REVIEW

**Comment by cptbtptpbcptdtptp** [comment]
Demo:
```
/**
 * @title Sprite Slice
 * @category 2D
 */
import * as dat from "dat.gui";
import {
  AssetType,
  Camera,
  Color,
  Entity,
  MathUtil,
  MeshRenderer,
  MeshTopology,
  ModelMesh,
  Script,
  Sprite,
  SpriteDrawMode,
  SpriteRenderer,
  SubMesh,
  Texture2D,
  UnlitMaterial,
  Vector4,
  WebGLEngine
} from "oasis-engine";

// Create engine object.
const engine = new WebGLEngine("canvas");
engine.canvas.resizeByClientSize();

// Create root entity.
const rootEntity = engine.sceneManager.activeScene.createRootEntity();

// Create camera.
const cameraEntity = rootEntity.createChild("Camera");
cameraEntity.transform.setPosition(0, 0, 5);
const camera = cameraEntity.addComponent(Camera);
camera.isOrthographic = true;
camera.orthographicSize = 5;

engine.resourceManager
  .load<Texture2D>({
    url: "https://gw.alipayobjects.com/mdn/rms_7c464e/afts/img/A*0vm_SJVssKAAAAAAAAAAAAAAARQnAQ",
    type: AssetType.Texture2D
  })
  .then((texture) => {
    // Create origin sprite entity.
    const spriteEntity = rootEntity.createChild("spriteSlice");
    const spriteRenderer = spriteEntity.addComponent(SpriteRenderer);
    const sprite = (spriteRenderer.sprite = new Sprite(engine, texture));
    spriteRenderer.drawMode = SpriteDrawMode.Sliced;
    sprite.border = new Vector4();
    addDataGUI(spriteEntity);
  });

engine.run();

/**
 * Auxiliary display.
 */
class TriangleScript extends Script {
  private triangleEntity: Entity;
  private modelMesh: ModelMesh;
  private targetSpriteRenderer: SpriteRenderer;

  onAwake(): void {
    const { engine, entity } = this;
    const triangleEntity = (this.triangleEntity = entity.createChild("tag"));
    const meshRenderer = triangleEntity.addComponent(MeshRenderer);
    meshRenderer.mesh = this.modelMesh = new ModelMesh(engine, "tag");
    const material = new UnlitMaterial(engine);
    material.baseColor = new Color(1, 0, 0, 1);
    meshRenderer.setMaterial(material);
    this.targetSpriteRenderer = entity.getComponent(SpriteRenderer);
  }

  setShow(value: boolean) {
    this.enabled = this.triangleEntity.isActive = value;
  }

  onUpdate(): void {
    const { modelMesh, targetSpriteRenderer } = this;
    // @ts-ignore
    const { positions, vertexCount, triangles } = targetSpriteRenderer._renderData;
    if (vertexCount > 0) {
      const length = triangles.length * 2;
      if (length <= 0) {
        return;
      }
      const myTriangles = modelMesh.getIndices() || new Uint16Array(length);
      const myPositions = modelMesh.getPositions() || [];
      for (let i = 0; i < positions.length; i++) {
        myPositions[i] = positions[i].clone();
      }
      for (let i = 0; i < triangles.length / 3; i++) {
        myTriangles[6 * i] = triangles[i * 3];
        myTriangles[6 * i + 1] = triangles[i * 3 + 1];
        myTriangles[6 * i + 2] = triangles[i * 3 + 1];
        myTriangles[6 * i + 3] = triangles[i * 3 + 2];
        myTriangles[6 * i + 4] = triangles[i * 3 + 2];
        myTriangles[6 * i + 5] = triangles[i * 3];
      }
      modelMesh.setPositions(myPositions);
      modelMesh.setIndices(myTriangles);
      modelMesh.clearSubMesh();
      modelMesh.addSubMesh(new SubMesh(0, length, MeshTopology.Lines));
      modelMesh.uploadData(false);
    }
  }
}

/**
 * Add data GUI.
 */
function addDataGUI(entity: Entity) {
  const spriteRenderer = entity.getComponent(SpriteRenderer);
  const triangleScript = entity.addComponent(TriangleScript);
  triangleScript.setShow(false);
  const sprite = spriteRenderer.sprite;
  const border = sprite.border;
  const gui = new dat.GUI();
  const defaultWidth = spriteRenderer.width;
  const defaultHeight = spriteRenderer.height;
  const guiData = {
    drawMode: "Slice",
    left: 0,
    bottom: 0,
    right: 0,
    top: 0,
    showTriangle: false,
    width: defaultWidth,
    height: defaultHeight,
    reset: () => {
      spriteRenderer.width = guiData.width = defaultWidth;
      spriteRenderer.height = guiData.height = defaultHeight;
      guiData.left = guiData.bottom = guiData.right = guiData.top = 0;
      sprite.border = border.set(0, 0, 0, 0);
      guiData.showTriangle = false;
    }
  };

  gui
    .add(guiData, "drawMode", ["Simple", "Slice", "Tiled"])
    .onChange((value: string) => {
      switch (value) {
        case "Simple":
          spriteRenderer.drawMode = SpriteDrawMode.Simple;
          break;
        case "Slice":
          spriteRenderer.drawMode = SpriteDrawMode.Sliced;
          break;
        case "Tiled":
          spriteRenderer.drawMode = 2;
          break;
        default:
          break;
      }
    })
    .listen();
  gui
    .add(guiData, "left", 0.0, 1.0, 0.01)
    .onChange((value: number) => {
      guiData.left = border.x = MathUtil.clamp(value, 0, 1 - border.z);
      sprite.border = border;
    })
    .listen();
  gui
    .add(guiData, "bottom", 0.0, 1.0, 0.01)
    .onChange((value: number) => {
      guiData.bottom = border.y = MathUtil.clamp(value, 0, 1 - border.w);
      sprite.border = border;
    })
    .listen();
  gui
    .add(guiData, "right", 0.0, 1.0, 0.01)
    .onChange((value: number) => {
      guiData.right = border.z = MathUtil.clamp(value, 0, 1 - border.x);
      sprite.border = border;
    })
    .listen();
  gui
    .add(guiData, "top", 0.0, 1.0, 0.01)
    .onChange((value: number) => {
      guiData.top = border.w = MathUtil.clamp(value, 0, 1 - border.y);
      sprite.border = border;
    })
    .listen();
  gui
    .add(guiData, "showTriangle")
    .onChange((value: boolean) => {
      triangleScript.setShow(value);
    })
    .listen();
  gui
    .add(guiData, "width", defaultWidth / 5, defaultWidth * 5, defaultWidth / 10)
    .onChange((value: number) => {
      spriteRenderer.width = value;
    })
    .listen();
  gui
    .add(guiData, "height", defaultHeight / 5, defaultHeight * 5, defaultHeight / 10)
    .onChange((value: number) => {
      spriteRenderer.height = value;
    })
    .listen();
  gui.add(guiData, "reset").name("重置");

  return guiData;
}

```

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 1]
```diff
@@ -77,6 +86,39 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
...
+    if (this._tileMode !== tileMode) {
+      this._tileMode = tileMode;
+      if (this.drawMode === SpriteDrawMode.Tiled) {
+        this._dirtyUpdateFlag |= SpriteRendererUpdateFlags.All;
+      }
+    }
+  }
+
+  /**
+   * The stretch value of the sprite renderer. (Only works in tiled mode.)
```
Comment (singlecoder) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:106]: Add range of values

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 2]
```diff
@@ -77,6 +86,39 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
...
+   */
+  get tileStretchValue(): number {
+    return this._tileStretchValue;
+  }
+
+  set tileStretchValue(stretchValue: number) {
+    stretchValue = MathUtil.clamp(stretchValue, 0, 1);
```
Comment (singlecoder) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:113]: Put this after if should be better

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]
Supplementary unit tests for new features

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    const spriteUVs = sprite._getUVs();
+    const spritePositions = sprite._getPositions();
+    const { x: left, y: bottom } = spritePositions[0];
+    const { x: right, y: top } = spritePositions[3];
+    const { width: expectWidth, height: expectHeight } = sprite;
+    const fixedLeft = expectWidth * border.x;
+    const fixedBottom = expectHeight * border.y;
+    const fixedRight = expectWidth * border.z;
+    const fixedTop = expectHeight * border.w;
+    if (fixedLeft + fixedRight >= width) {
```
Comment (singlecoder) [inline packages/core/src/2d/assembler/TiledSpriteAssembler.ts:214]: The result of fixedLeft + fixedRight used many times，can be written as follows
```typescript
const fixedWidth = fixedLeft + fixedRight;
if (fixedWidth >= width) {
} else {
}
```

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      } else {
+        posRow.push(width * left, width * right);
+        posColumn.push(height * bottom, height * top);
+        uvRow.push(spriteUVs[0].x, spriteUVs[3].x);
+        uvColumn.push(spriteUVs[0].y, spriteUVs[3].y);
+        return;
+      }
+    }
+
+    if (fixedTop + fixedBottom >= height) {
```
Comment (singlecoder) [inline packages/core/src/2d/assembler/TiledSpriteAssembler.ts:251]: The same as fixedLeft + fixedRight

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+      if (count === 0) {
+        posRow.push(expectWidth * left, fixedLeft, width - fixedRight, width - expectWidth * (1 - right));
+        uvRow.push(spriteUVs[0].x, spriteUVs[1].x, NaN, NaN, spriteUVs[2].x, spriteUVs[3].x);
+      } else if (count <= Basic2DBatcher.MAX_VERTEX_COUNT / 4 / 4 - 4) {
+        posRow.push(expectWidth * left, fixedLeft);
+        uvRow.push(spriteUVs[0].x, spriteUVs[1].x, spriteUVs[1].x);
+        const countInteger = count | 0;
+        for (let i = 0; i < countInteger; i++) {
+          posRow.push(fixedLeft + (i + 1) * fixedCenterW);
+          uvRow.push(spriteUVs[2].x, spriteUVs[1].x);
```
Comment (singlecoder) [inline packages/core/src/2d/assembler/TiledSpriteAssembler.ts:234]: spriteUVs[1] and spriteUVs[2] repeat many times.

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+      );
+      uvRow.push(spriteUVs[0].x, spriteUVs[1].x, spriteUVs[2].x, spriteUVs[3].x);
+    } else {
+      const fixedCenterW = expectWidth - fixedLeft - fixedRight;
+      const count = fixedCenterW > zeroTolerance ? (width - fixedLeft - fixedRight) / fixedCenterW : 0;
+      if (count === 0) {
+        posRow.push(expectWidth * left, fixedLeft, width - fixedRight, width - expectWidth * (1 - right));
+        uvRow.push(spriteUVs[0].x, spriteUVs[1].x, NaN, NaN, spriteUVs[2].x, spriteUVs[3].x);
+      } else if (count <= Basic2DBatcher.MAX_VERTEX_COUNT / 4 / 4 - 4) {
+        posRow.push(expectWidth * left, fixedLeft);
```
Comment (singlecoder) [inline packages/core/src/2d/assembler/TiledSpriteAssembler.ts:229]: All SpriteRenderer object with tiled draw mode will use Basic2DBatcher.MAX_VERTEX_COUNT / 4 / 4，should be static, and Basic2DBatcher.MAX_VERTEX_COUNT / 16 is better

Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/assembler/TiledSpriteAssembler.ts:229]: Done.

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts (hunk 4 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/assembler/TiledSpriteAssembler.ts:251]: Done.

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts (hunk 3 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/assembler/TiledSpriteAssembler.ts:214]: Done.

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 2 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/SpriteRenderer.ts:113]: Done.

**Comment by cptbtptpbcptdtptp** [comment]
> 

Done.

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 7]
```diff
@@ -77,6 +85,39 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
...
+      if (this.drawMode === SpriteDrawMode.Tiled) {
+        this._dirtyUpdateFlag |= SpriteRendererUpdateFlags.All;
+      }
+    }
+  }
+
+  /**
+   * The stretch value of the sprite renderer, specified in normalized. (Only works in tiled mode.)
+   */
+  get tileStretchValue(): number {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:107]: Only work in tile adaptive mode

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 8]
```diff
@@ -215,6 +256,9 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
    * @internal
    */
   _cloneTo(target: SpriteRenderer): void {
+    target._tileMode = this._tileMode;
+    target._tileStretchValue = this._tileStretchValue;
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:260]: Why no use @assignmentClone for `_tileMode` and `_tileStretchValue`

File: packages/core/src/2d/enums/SpriteTileMode.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+ * Sprite's tiling mode enumeration.
+ * Only works in `SpriteDrawMode.tiled` mode.
+ */
+export enum SpriteTileMode {
+  /** SpriteRenderer will tile continuously. */
+  Continuous,
+  /** When the size of the intermediate expansion is greater than or equal to `SpriteRenderer._tileStretchValue`, 
+   * a tile will be added，and the number of tiles can only be an integer. */ 
+  Adaptive
+}
```
Comment (GuoLei1990) [inline packages/core/src/2d/enums/SpriteTileMode.ts:11]: /**
 * Tiling mode when the sprite drawing mode is tiled.
 */
export enum SpriteTileMode {
  /** Continuously tile sprites while rendering. */
  Continuous,
  
  /**
   * Adaptive tiling when rendering.
   * 
   * @remarks
   * When the size of the intermediate expansion is greater than or equal to `SpriteRenderer.tileStretchValue`, 
   * a tile will be added，and the number of tiles can only be an integer. 
   */
  Adaptive
}

File: packages/core/src/2d/enums/SpriteDrawMode.ts [hunk 10]
```diff
@@ -4,6 +4,8 @@
 export enum SpriteDrawMode {
   /** Overall scaling when modifying size. */
   Simple,
-  /** When modifying the size, it is transformed according to the 9-slice settings (border). */
-  Sliced
+  /** When modifying the size of the renderer, it scales to fill the range according to the 9-slice settings (border). */
+  Sliced,
+  /** When modifying the size of the renderer, it will tile to fill the range according to the 9-slice settings (border). */
+  Tiled
```
Comment (GuoLei1990) [inline packages/core/src/2d/enums/SpriteDrawMode.ts:10]: Wrong comments

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 11]
```diff
@@ -77,6 +85,39 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
...
+  }
+
+  /**
+   * The stretch value of the sprite renderer, specified in normalized. (Only works in tiled mode.)
+   */
+  get tileStretchValue(): number {
+    return this._tileStretchValue;
+  }
+
+  set tileStretchValue(stretchValue: number) {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:111]: tileAdaptiveStretchThreshold

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 7 above)
Comment (GuoLei1990) [reply packages/core/src/2d/sprite/SpriteRenderer.ts:107]: Tile Adaptive Threshold

**Review by GuoLei1990 — APPROVED** [review]

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts (hunk 5 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/assembler/TiledSpriteAssembler.ts]: Done.

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+    const rowLength = posRow.length - 1;
+    const columnLength = posColumn.length - 1;
+    let count = 0;
+    let trianglesOffset = 0;
+    for (let j = 0; j < columnLength; j++) {
+      for (let i = 0; i < rowLength; i++) {
+        const uvL = uvRow.get(2 * i);
+        const uvB = uvColumn.get(2 * j);
+        const uvR = uvRow.get(2 * i + 1);
+        const uvT = uvColumn.get(2 * j + 1);
```
Comment (singlecoder) [inline packages/core/src/2d/assembler/TiledSpriteAssembler.ts:64]: 2 * i and 2 * j repeat

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+
+        // right and bottom
+        uvs[count] ? uvs[count].set(uvR, uvB) : (uvs[count] = new Vector2(uvR, uvB));
+        pos = positions[count];
+        if (pos) {
+          pos.set(wE0 * r + wE4 * b + wE12, wE1 * r + wE5 * b + wE13, wE2 * r + wE6 * b + wE14);
+        } else {
+          positions[count] = new Vector3(wE0 * r + wE4 * b + wE12, wE1 * r + wE5 * b + wE13, wE2 * r + wE6 * b + wE14);
+        }
+        count++;
```
Comment (singlecoder) [inline packages/core/src/2d/assembler/TiledSpriteAssembler.ts:97]: wE4 * b repeat

File: packages/core/src/2d/assembler/TiledSpriteAssembler.ts (hunk 12 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/assembler/TiledSpriteAssembler.ts]: I saved more than two calculations as temporary variables.

**Review by singlecoder — APPROVED** [review]
+1

**Pull request merged**